#Task1

_References_
[Calculating Decision tree using Hunts Algorithm](http://www.saedsayad.com/decision_tree.htm), 
[Gini index](https://stats.stackexchange.com/questions/77213/computing-the-gini-index)

Consider the data given in Table 1. Construct a decision tree using the following rules: (i) at each step
compute the gini index for every possible split considering all attributes and select the split with the best
gini value; (ii) Stopping rule: when the **gini value of a node is zero** or **no further split is possible**; and (iii)
the class of a leaf node is determined by a majority rule (breaking ties arbitrarily). Using the decision tree
you built, predict the class value of the record: ('Rainy','Hot',Windy='F').



Outlook Temperature Windy Play  
Sunny Hot False Y  
Sunny Mild False Y  
Rainy Mild False N  
Sunny Hot True N  
Sunny Hot True N  
Rainy Mild True N  

In [44]:
import math
def gini(node):
    ans=1.0
    sum=0.0
    for c in node:
        sum+=c
    for c in node:
        ans-=math.pow(c/sum,2)
        #print(math.pow(c/sum,2))
        #print(ans)
    return ans

def summation(arr):
    sum=0
    for n in arr:
        sum+=n
    return sum

def giniCalculation(feature, n1,n2):
    print(feature)
    print("Y N")
    g1=gini(n1)
    g2=gini(n2)
    total=summation(n1)+summation(n2)
    tGini=summation(n1)/total*g1+summation(n2)/total*g2

    print("n1",n1)
    print("n2",n2)
    print("g1",g1)
    print("g2",g2)
    print("total",total)
    print("tGini", tGini)

##Choosing the first Split 
###Outlook
Outlook Temperature Windy Play  
Sunny Hot False Y  
Sunny Mild False Y  
Sunny Hot True N  
Sunny Hot True N  


Outlook Temperature Windy Play  
Rainy Mild False N  
Rainy Mild True N  




In [45]:
giniCalculation("Outlook", [2.0,2.0],[0.0,2.0])

Outlook
Y N
('n1', [2.0, 2.0])
('n2', [0.0, 2.0])
('g1', 0.5)
('g2', 0.0)
('total', 6.0)
('tGini', 0.3333333333333333)


###Temperature

Outlook Temperature Windy Play  
Sunny Hot False Y  
Sunny Hot True N  
Sunny Hot True N  

Rainy Mild True N      
Sunny Mild False Y      
Rainy Mild False N   

In [46]:
giniCalculation("Temperature",[1.0,2.0],[1.0,2.0])

Temperature
Y N
('n1', [1.0, 2.0])
('n2', [1.0, 2.0])
('g1', 0.4444444444444444)
('g2', 0.4444444444444444)
('total', 6.0)
('tGini', 0.4444444444444444)


###Windy

Outlook Temperature Windy Play  
Sunny Hot True N  
Sunny Hot True N  
Rainy Mild True N      

Sunny Hot False Y  
Sunny Mild False Y      
Rainy Mild False N   

In [47]:
giniCalculation("Windy",[0.0,3.0],[2.0,1.0])

Windy
Y N
('n1', [0.0, 3.0])
('n2', [2.0, 1.0])
('g1', 0.0)
('g2', 0.4444444444444445)
('total', 6.0)
('tGini', 0.22222222222222224)


Choosing Windy as Split attribute, because it has the lowest Gini Index.   
**Node1**('g1', 0.0)Stopping condiditon reached at Node1, because gini = 0.0  
Outlook Temperature Windy Play  
Sunny Hot True N  
Sunny Hot True N  
Rainy Mild True N      
**Node2**('g2', 0.4444444444444445)    
Sunny Hot False Y  
Sunny Mild False Y      
Rainy Mild False N   

## Choosing the second split on Node 2

### Outlook 
Outlook Temperature Windy Play    
Sunny Hot False Y  
Sunny Mild False Y  

      
Rainy Mild False N   




In [48]:
giniCalculation("Outlook", [2.0,0.0],[0.0,1.0])

Outlook
Y N
('n1', [2.0, 0.0])
('n2', [0.0, 1.0])
('g1', 0.0)
('g2', 0.0)
('total', 3.0)
('tGini', 0.0)


### Temperature

Outlook Temperature Windy Play  
Sunny Hot False Y  

Sunny Mild False Y  
Rainy Mild False N   




In [49]:
giniCalculation("Temperature",[1.0,0.0],[1.0,1.0])

Temperature
Y N
('n1', [1.0, 0.0])
('n2', [1.0, 1.0])
('g1', 0.0)
('g2', 0.5)
('total', 3.0)
('tGini', 0.3333333333333333)


Choosing Outlook as a split attribute, because it has the lowest gini index, 0.

### Outlook 

**Node1** ('g1', 0.0) Stopping condition reached   
Outlook Temperature Windy Play    
Sunny Hot False Y  
Sunny Mild False Y  

**Node** ('g2', 0.0) Stopping condition reached    
Rainy Mild False N   


Since we have reached stopping condition on all the subtrees, we will stop. 

